accession/accession.py

```diff
@@ -564,18 +571,220 @@ def accession_steps(self):
+        qc_file = self.analysis.get_files(
+            filename=gs_file.task.outputs[qc_file_task_output_name]
+        )[0]
+        qc = self.backend.read_json(qc_file)
+        output_qc = qc.get(qc_dictionary_key)
+        return self.queue_qc(output_qc, encode_file, qc_schema_name_with_hyphens)
+
+    def make_flagstat_qc(
+        self,
+        encode_file: Dict[str, Any],
+        gs_file: GSFile,
+        task_output_name: str,
+        qc_dictionary_key: str,
+        convert_to_string: List = [
+            "mapped_pct",
+            "paired_properly_pct",
+            "singletons_pct",
+        ],
+    ):
+        if self.file_has_qc(encode_file, "SamtoolsFlagstatsQualityMetric"):
+            return
+        qc_file = self.analysis.get_files(
+            filename=gs_file.task.outputs[task_output_name]
+        )[0]
+        qc = self.backend.read_json(qc_file)
+        output_qc = qc.get(qc_dictionary_key)
+        for key in convert_to_string:
+            # paired_properly_pct and singletons_pct are not there in single-ended
+            try:
+                output_qc[key] = str(output_qc[key])
+            except KeyError:
+                continue
+
+        qc_bytes = self.get_bytes_from_dict(qc)
+        attachment = self.make_attachment_object(
+            qc_bytes, "text/plain", qc_file.filename, ".txt"
+        )
+        output_qc["attachment"] = attachment
+        return self.queue_qc(
+            output_qc, encode_file, "samtools-flagstats-quality-metric"
+        )
+
+    def make_genome_flagstat_qc(self, encode_file: Dict[str, Any], gs_file: GSFile):
+        self.make_flagstat_qc(
+            encode_file, gs_file, "genome_flagstat_json", "samtools_genome_flagstat"
+        )
+
+    def make_anno_flagstat_qc(self, encode_file: Dict[str, Any], gs_file: GSFile):
+        self.make_flagstat_qc(
+            encode_file, gs_file, "anno_flagstat_json", "samtools_anno_flagstat"
+        )
+
+    def make_number_of_genes_detected_qc(
+        self, encode_file: Dict[str, Any], gs_file: GSFile
+    ):
+        self.make_qc_from_well_formed_json(
+            encode_file,
+            gs_file,
+            "GeneQuantificationQualityMetric",
+            "number_of_genes",
+            "number_of_genes_detected",
+            "gene-quantification-quality-metric",
+        )
+
+    def make_mad_qc_metric(self, encode_file: Dict[str, Any], gs_file: GSFile):
+        self.make_generic_correlation_qc(
+            encode_file,
+            gs_file,
+            self.prepare_mad_qc_metric,
+            "MadQualityMetric",
+            "mad-quality-metric",
+        )
+
+    def prepare_mad_qc_metric(self, gs_file: GSFile) -> Dict:
+        qc_file = self.analysis.search_down(gs_file.task, "mad_qc", "madQCmetrics")[0]
+        qc = self.backend.read_json(qc_file)
+        try:
+            mad_qc = qc["MAD.R"]
+        except KeyError:
+            self.logger.exception("Something is wrong with the madqc source file")
```

Comment:
Should be explicit about the missing key in the traceback:
```suggestion
        try:
            qc_key = "MAD.R"
            mad_qc = qc[qc_key]
        except KeyError:
            self.logger.exception("Could not find key %s in madqc source file", qc_key)
```